[1ES] Update extensions build template

liliankasem opened the pull request:
<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Update extensions build template to allow multiple projects to be build and packed.

jviau commented on eng/ci/templates/official/jobs/build-extension-artifacts.yml:
```diff
@@ -1,10 +1,13 @@
 jobs:
 
-- job: BuildArtifacts
+- job: ${{ parameters.JobName }}
```
You could also calculate the job name from `ExtensionProjectName`

```suggestion
- job: Build_${{ parameters.ExtensionProjectName }}
```

liliankasem replied:
I don't think `.` are supported in job names?

> name of the job, A-Z, a-z, 0-9, and underscore

liliankasem replied:
Taking a different approach, this wont be required

liliankasem commented:
@jviau - I changed the approach here due to a number of limitations - mind re-reviewing?

jviau commented:
> @jviau - I changed the approach here due to a number of limitations - mind re-reviewing?

Which limitations where you encountering?

jviau commented:
An alternative is instead of calling the template multiple times, you could provide multiple projects into the parameter and then for-each add tasks to build them.

``` yml
parameters:
- name: projects
  type: object

.
.
.

    steps:
    - ${{ each project in parameters.projects }}:
      -  template: build-extension.yml
         parameters:
           - project: ${{ project ""
```

``` yml
# build-extension.yml
parameters:
- name: project
  type: string

steps:
  - task: DotNetCoreCli@2
    ....
```

liliankasem commented on eng/ci/templates/official/jobs/build-extension-artifacts.yml:
```diff
@@ -80,7 +80,7 @@ jobs:
       arguments: '--no-build -c release -o $(Build.ArtifactStagingDirectory)/NugetPackages -p:BuildNumber=$(buildNumber)'
       projects: |
-        extensions/${{ parameters.ExtensionProjectName }}/src/*.csproj
+        extensions/${{ parameters.ExtensionProjectName }}*/**/*.csproj
```
It's intentional - this is the solution to the issues I'm seeing (see other comment)

liliankasem commented:
> > @jviau - I changed the approach here due to a number of limitations - mind re-reviewing?
> 
> Which limitations where you encountering?

The main issue is the NuGet Publish task: there is a conflict error when we publish twice, so I refactored the template to set a publish flag and so the CI would only publish on the second run (using dependsOn). But then I realized this doesn't work as the folders would only contain the one nuget package from the job it was published from. An easy solution would be to have a different name for the nuget package folder that we publish, but I don't like this idea. I would like to keep the naming the same and consistent across pipelines so that release pipelines only need to look for one folder instead of having it look through multiple folders in a given CI run.

After all the back and fourth, I landed on using `extensions/${{ parameters.ExtensionProjectName }}*/**/*.csproj`, this would allow the pipeline to pick up any extension folder that starts with the project name such as:

- `extensions/Worker.Extensions.Http.AspNetCore` and
- `extensions/Worker.Extensions.Http.AspNetCore.Analyzers`

However, I do like your proposed suggestion of using an object and looping through to run `pack` for each project via `${{ each project in parameters.projects }}:`

Both solve the issue, but the latter lets you be more explicit about which projects are packaged and I prefer that a lot more :)

jviau commented on eng/ci/templates/official/jobs/build-extension-artifacts.yml:
```diff
@@ -21,93 +25,20 @@ jobs:
-      arguments: -p:BuildNumber=$(buildNumber) -c release
-      projects: |
-        extensions/${{ parameters.ExtensionProjectName }}/src/*.csproj
+  - ${{ if ne(parameters.ExtensionProjectName, '') }}:
```
I think I would still rather have just 1 way of providing the projects here.

liliankasem replied:
😞

liliankasem replied:
I think I need to update them all so will think on this in my next PR
